libraries/AP_Math/vector3.cpp

```diff
@@ -474,9 +483,89 @@ float Vector3<T>::closest_distance_between_line_and_point(const Vector3<T> &w1,
     float dot_product = unit_vec * scaled_p_vec;
     dot_product = constrain_float(dot_product,0.0f,1.0f); 
  
-    const Vector3<T> nearest = line_vec * dot_product;
-    const float dist = (nearest - p_vec).length();
-    return dist;
+    const Vector3<T> closest_point = line_vec * dot_product;
+    return (closest_point + w1);
+}
+
+// This implementation is borrowed from: http://geomalgorithms.com/a07-_distance.html
+// INPUT: 4 points corresponding to start and end of two line segments
+// OUTPUT: shortest distance between segments, and closest point on segment 2, from segment 1, gets passed on reference as "intersection" 
+template <typename T>
+float Vector3<T>::segment_to_segment_dist(const Vector3<T>& seg1_start, const Vector3<T>& seg1_end, const Vector3<T>& seg2_start, const Vector3<T>& seg2_end, Vector3<T>& intersection)
```

Comment:
These changes require updates to the math test suite.

Reply:
Still missing these tests.  They're really easy to add - the tests are in `libraries/AP_Math/tests/`

You run them like this (e.g.):

```
./waf configure --board=linux --debug
./waf tests
./build/linux/tests/test_vector2
```

I was going to point you to our Wiki page on the `Unit Test` framework but discovered we didn't have one.

So I'm creating it with this PR: https://github.com/ArduPilot/ardupilot_wiki/pull/3251/files